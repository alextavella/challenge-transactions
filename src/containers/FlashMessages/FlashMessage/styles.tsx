import { Container, Typography } from '@material-ui/core';
import { animated } from 'react-spring';
import styled, { css } from 'styled-components';

interface FlashItemProps {
  type: 'error' | 'success';
}

const styles = {
  error: css`
    background-color: #e5433d;
    color: #fff;
  `,
  success: css`
    background-color: #80bf66;
    color: #fff;
  `,
};

export const FlashItem = styled(animated.li)<FlashItemProps>`
  ${props => styles[props.type]};

  cursor: pointer;
  padding: 16px;
  position: relative;
`;

export const Content = styled(Container).attrs({
  maxWidth: 'sm',
})`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;

  svg {
    margin-right: 8px;
  }
`;

export const TitleText = styled(Typography).attrs({
  variant: 'body1',
  component: 'span',
})``;
