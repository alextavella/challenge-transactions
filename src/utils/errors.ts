import { ValidationError } from 'yup';

interface Error {
  [key: string]: string;
}

export function getValidationErrors(errors: ValidationError): any {
  return errors.inner.reduce((acc: Error, error: ValidationError) => {
    return { ...acc, [error.path]: error.message };
  }, {});
}
