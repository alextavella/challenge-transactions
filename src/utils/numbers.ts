export const formatCurrency = (value: number): string => {
  const formatted = new Intl.NumberFormat([], {
    style: 'currency',
    currency: 'BRL',
    minimumFractionDigits: 2,
  }).format(value);

  return formatted
    .replace('.', '#')
    .replace(/,/g, '.')
    .replace('#', ',')
    .replace('R$', 'R$ ');
};

export const unformatCurrency = (value: string): number => {
  return Number(
    value
      .replace('R$ ', '')
      .replace(',', '#')
      .replace(/\./g, '')
      .replace('#', '.'),
  );
};
