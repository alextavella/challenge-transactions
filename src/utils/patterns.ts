import { ValidationRules } from 'react-hook-form';

export interface MaskRule {
  rules: ValidationRules;
  transform: (text: string) => string;
}

export const CPF: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{3}).([0-9]{3}).([0-9]{3})-([0-9]{2})$/),
      message: 'CPF inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,3})(\d{0,3})(\d{0,3})(\d{0,2})/)!;

    const part1 = result[2] ? `${result[1]}.${result[2]}` : result[1];
    const part2 = result[3] ? `.${result[3]}` : '';
    const part3 = result[4] ? `-${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_NUMBER: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^([0-9]{4})\s([0-9]{4})\s([0-9]{4})\s([0-9]{4})$/),
      message: 'Número do cartão inválido',
    },
  },
  transform: (text: string = ''): string => {
    const result = text
      .replace(/\D/g, '')
      .match(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/)!;

    const part1 = result[2] ? `${result[1]} ${result[2]}` : result[1];
    const part2 = result[3] ? ` ${result[3]}` : '';
    const part3 = result[4] ? ` ${result[4]}` : '';

    return `${part1}${part2}${part3}`;
  },
};

export const CREDIT_CARD_EXPIRATION_DATE: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/^(0[1-9]|1[012])\/?(20[2-9][0-9])$/),
      message: 'Data de expiração inválida',
    },
  },
  transform: (text: string = ''): string => {
    const result = text.replace(/\D/g, '').match(/(\d{0,2})(\d{0,4})/)!;

    return result[2] ? `${result[1]}/${result[2]}` : result[1];
  },
};

export const CREDIT_CARD_CVV: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/[0-9]{3}/),
      message: 'CVV inválido',
    },
  },
  transform: (text: string = ''): string => {
    return text.substr(0, 3);
  },
};

export const CURRENCY: MaskRule = {
  rules: {
    pattern: {
      value: new RegExp(/([0-9]\.?)?[0-9]+,[0-9]{2}/),
      message: 'Valor inválido',
    },
  },
  transform: (text: string = ''): string => {
    let number = text.replace(/[^0-9]/g, '');

    if (number.length === 0) number = '0.00';
    else if (number.length === 1) number = '0.0' + number;
    else if (number.length === 2) number = '0.' + number;
    else
      number =
        number.substring(0, number.length - 2) +
        '.' +
        number.substring(number.length - 2, number.length);

    number = Number(number).toFixed(2);
    number = number.replace(/\./g, ',');

    const x = number.split(',');
    const x1 = x[0];
    const x2 = x.length > 1 ? ',' + x[1] : '';

    const formatAmountNoDecimals = (number: string) => {
      var rgx = /(\d+)(\d{3})/;
      while (rgx.test(number)) {
        number = number.replace(rgx, `$1.$2`);
      }
      return number;
    };

    return `R$ ` + formatAmountNoDecimals(x1) + x2;
  },
};
