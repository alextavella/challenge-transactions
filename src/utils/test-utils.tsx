import { render } from '@testing-library/react';
import React from 'react';
import MainProvider from '../providers';

const AllTheProviders: React.FC = ({ children }) => {
  return <MainProvider>{children}</MainProvider>;
};

const customRender = (ui: React.ReactElement, options?: any) =>
  render(ui, {
    wrapper: AllTheProviders,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
