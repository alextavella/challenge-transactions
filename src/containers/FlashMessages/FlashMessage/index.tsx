import SuccessIcon from '@material-ui/icons/CheckCircle';
import ErrorIcon from '@material-ui/icons/Error';
import React from 'react';
import { Content, FlashItem, TitleText } from './styles';

export interface FlashMessageProps {
  type: 'error' | 'success';
  title: string;
  onClick(): void;
  style?: Object;
}

const icons = {
  error: ErrorIcon,
  success: SuccessIcon,
};

const FlashMessage: React.FC<FlashMessageProps> = ({
  title,
  type,
  onClick: click,
  ...rest
}) => {
  const Icon: React.ComponentType = icons[type];

  return (
    <FlashItem type={type} onClick={click} {...rest}>
      <Content>
        <Icon />
        <TitleText>{title}</TitleText>
      </Content>
    </FlashItem>
  );
};

export default FlashMessage;
