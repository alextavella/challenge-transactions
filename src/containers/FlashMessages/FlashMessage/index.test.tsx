import React from 'react';
import FlashMessage from '.';
import { cleanup, fireEvent, render, screen } from '../../../utils/test-utils';

describe('<FlashMessage />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should be able to render correctly when error type', () => {
    // Arrange
    const { baseElement } = render(
      <FlashMessage type="error" title="Message" onClick={jest.fn()} />,
    );

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to render correctly when success type', () => {
    // Arrange
    const { baseElement } = render(
      <FlashMessage type="success" title="Message" onClick={jest.fn()} />,
    );

    // Assert
    expect(baseElement).toMatchSnapshot();
  });

  it('should be able to click on component', () => {
    // Arrange
    const spyClick = jest.fn();

    render(<FlashMessage type="success" title="click-me" onClick={spyClick} />);

    // Act
    const message = screen.getByText('click-me');
    fireEvent.click(message);

    // Assert
    expect(spyClick).toHaveBeenCalled();
  });
});
