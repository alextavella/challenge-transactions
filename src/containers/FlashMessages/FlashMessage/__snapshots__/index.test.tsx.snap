// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlashMessage /> should be able to render correctly when error type 1`] = `
<body>
  <div>
    <li
      class="sc-bdfBwQ iHLimE"
      type="error"
    >
      <div
        class="MuiContainer-root sc-gsTCUz kXSENp MuiContainer-maxWidthSm"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
        <h6
          class="MuiTypography-root sc-dlfnbm MuiTypography-h6"
        >
          Message
        </h6>
      </div>
    </li>
  </div>
</body>
`;

exports[`<FlashMessage /> should be able to render correctly when success type 1`] = `
<body>
  <div>
    <li
      class="sc-bdfBwQ iTuIRM"
      type="success"
    >
      <div
        class="MuiContainer-root sc-gsTCUz kXSENp MuiContainer-maxWidthSm"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <h6
          class="MuiTypography-root sc-dlfnbm MuiTypography-h6"
        >
          Message
        </h6>
      </div>
    </li>
  </div>
</body>
`;
